File: php/tests/BankIdSeTest.php
<?php

declare(strict_types=1);

namespace Idkollen\Client\Tests;

use GuzzleHttp\Psr7\HttpFactory;
use Idkollen\Client\Endpoints\BankIdSeEndpoint;
use Idkollen\Client\Models\BankIdSe\BankIdSeAuthRequest;
use Idkollen\Client\Models\BankIdSe\BankIdSeCompleted;
use Idkollen\Client\Models\BankIdSe\BankIdSeFailed;
use Idkollen\Client\Models\BankIdSe\BankIdSePending;
use Idkollen\Client\Models\BankIdSe\BankIdSePendingPhone;
use Idkollen\Client\Models\BankIdSe\BankIdSePhoneAuthRequest;
use Idkollen\Client\Models\BankIdSe\BankIdSeVerifyRequest;
use Idkollen\Client\Transport;
use PHPUnit\Framework\TestCase;

final class BankIdSeTest extends TestCase
{
    private MockHttpClient $http;
    private BankIdSeEndpoint $endpoint;

    protected function setUp(): void
    {
        $this->http = new MockHttpClient();
        $factory = new HttpFactory();
        $transport = new Transport($this->http, $factory, $factory, 'https://x.test', 'cid', 'sec');
        $this->endpoint = new BankIdSeEndpoint($transport);
    }

    public function testAuthPending(): void
    {
        $this->http->enqueue(200, [
            'status' => 'PENDING',
            'id' => 'abc',
            'refId' => null,
            'autoStartToken' => 'tok',
        ]);

        $result = $this->endpoint->auth(new BankIdSeAuthRequest(ssn: '199001011234'));

        self::assertInstanceOf(BankIdSePending::class, $result);
        self::assertSame('abc', $result->id);
        self::assertSame('tok', $result->autoStartToken);

        $body = (string) $this->http->lastRequest()->getBody();
        self::assertSame(['ssn' => '199001011234'], json_decode($body, true));
    }

    public function testAuthCompleted(): void
    {
        $this->http->enqueue(200, [
            'status' => 'COMPLETED',
            'id' => 'abc',
            'ssn' => '199001011234',
            'name' => 'Test User',
            'givenName' => 'Test',
            'surname' => 'User',
        ]);

        $result = $this->endpoint->auth(new BankIdSeAuthRequest());

        self::assertInstanceOf(BankIdSeCompleted::class, $result);
        self::assertSame('Test User', $result->name);
    }

    public function testAuthFailed(): void
    {
        $this->http->enqueue(200, [
            'status' => 'FAILED',
            'id' => 'abc',
            'error' => 'userCancel',
        ]);

        $result = $this->endpoint->auth(new BankIdSeAuthRequest());

        self::assertInstanceOf(BankIdSeFailed::class, $result);
        self::assertSame('userCancel', $result->error);
    }

    public function testPhoneAuthPending(): void
    {
        $this->http->enqueue(200, [
            'status' => 'PENDING',
            'id' => 'abc',
            'hintCode' => 'outstandingTransaction',
        ]);

        $result = $this->endpoint->phoneAuth(new BankIdSePhoneAuthRequest(
            ssn: '199001011234',
            callInitiator: 'user',
        ));

        self::assertInstanceOf(BankIdSePendingPhone::class, $result);
    }

    public function testVerify(): void
    {
        $this->http->enqueue(200, [
            'ssn' => '199001011234',
            'name' => 'Test User',
            'givenName' => 'Test',
            'surname' => 'User',
            'age' => 34,
            'verifiedAt' => '2026-06-08T10:00:00Z',
        ]);

        $result = $this->endpoint->verify(new BankIdSeVerifyRequest(qrCode: 'qr'));

        self::assertSame(34, $result->age);
        self::assertSame('Test User', $result->name);
    }

    public function testCancelAuthSendsDelete(): void
    {
        $this->http->enqueue(204, '');
        $this->endpoint->cancelAuth('xyz');
        self::assertSame('DELETE', $this->http->lastRequest()->getMethod());
        self::assertSame('/v3/bankid-se/auth/xyz', $this->http->lastRequest()->getUri()->getPath());
    }

    public function testAuthStatusGet(): void
    {
        $this->http->enqueue(200, [
            'status' => 'PENDING',
            'id' => 'xyz',
        ]);
        $this->endpoint->authStatus('xyz');
        self::assertSame('GET', $this->http->lastRequest()->getMethod());
    }
}

File: go/idkollen/bankid_se_test.go
package idkollen_test

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	idkollen "github.com/idkollen/idk-clients/go/idkollen"
)

func TestBankIdSeAuth_Pending(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost || r.URL.Path != "/v3/bankid-se/auth" {
			t.Errorf("unexpected %s %s", r.Method, r.URL.Path)
		}
		_, _ = w.Write([]byte(`{"status":"PENDING","id":"abc","autoStartToken":"tok","qrStartToken":"qt","qrStartSecret":"qs"}`))
	}))
	defer srv.Close()

	c := idkollen.NewClientBuilder("id", "secret").BaseURL(srv.URL).Build()
	status, err := c.BankIdSe().Auth(context.Background(), idkollen.BankIdSeAuthRequest{})
	if err != nil {
		t.Fatal(err)
	}
	p, ok := status.(*idkollen.BankIdSePending)
	if !ok {
		t.Fatalf("expected *BankIdSePending, got %T", status)
	}
	if p.Id != "abc" {
		t.Errorf("expected id abc, got %s", p.Id)
	}
}

func TestBankIdSeAuth_Completed(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_, _ = w.Write([]byte(`{"status":"COMPLETED","id":"abc","ssn":"190001011234","name":"Test User","givenName":"Test","surname":"User"}`))
	}))
	defer srv.Close()

	c := idkollen.NewClientBuilder("id", "secret").BaseURL(srv.URL).Build()
	status, err := c.BankIdSe().Auth(context.Background(), idkollen.BankIdSeAuthRequest{})
	if err != nil {
		t.Fatal(err)
	}
	comp, ok := status.(*idkollen.BankIdSeCompleted)
	if !ok {
		t.Fatalf("expected *BankIdSeCompleted, got %T", status)
	}
	if comp.Ssn != "190001011234" {
		t.Errorf("expected ssn, got %s", comp.Ssn)
	}
}

func TestBankIdSePhoneAuth_PendingPhone(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v3/bankid-se/phone/auth" {
			t.Errorf("unexpected path: %s", r.URL.Path)
		}
		_, _ = w.Write([]byte(`{"status":"PENDING","id":"phone1","hintCode":"outstandingTransaction"}`))
	}))
	defer srv.Close()

	c := idkollen.NewClientBuilder("id", "secret").BaseURL(srv.URL).Build()
	status, err := c.BankIdSe().PhoneAuth(context.Background(), idkollen.BankIdSePhoneAuthRequest{Ssn: "190001011234", CallInitiator: "USER"})
	if err != nil {
		t.Fatal(err)
	}
	p, ok := status.(*idkollen.BankIdSePendingPhone)
	if !ok {
		t.Fatalf("expected *BankIdSePendingPhone, got %T", status)
	}
	if p.Id != "phone1" {
		t.Errorf("expected id phone1, got %s", p.Id)
	}
}

func TestBankIdSeVerify(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		var body map[string]any
		_ = json.NewDecoder(r.Body).Decode(&body)
		if body["qrCode"] == nil {
			t.Error("expected qrCode in body")
		}
		_, _ = w.Write([]byte(`{"ssn":"190001011234","name":"Test","givenName":"Test","surname":"User"}`))
	}))
	defer srv.Close()

	c := idkollen.NewClientBuilder("id", "secret").BaseURL(srv.URL).Build()
	resp, err := c.BankIdSe().Verify(context.Background(), idkollen.BankIdSeVerifyRequest{QrCode: "qrdata"})
	if err != nil {
		t.Fatal(err)
	}
	if resp.Ssn != "190001011234" {
		t.Errorf("expected ssn, got %s", resp.Ssn)
	}
}

func TestBankIdSeCancelAuth(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodDelete {
			t.Errorf("expected DELETE, got %s", r.Method)
		}
		w.WriteHeader(http.StatusNoContent)
	}))
	defer srv.Close()

	c := idkollen.NewClientBuilder("id", "secret").BaseURL(srv.URL).Build()
	if err := c.BankIdSe().CancelAuth(context.Background(), "abc"); err != nil {
		t.Fatal(err)
	}
}

func TestBankIdSeWaitForAuth_CompletesImmediately(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_, _ = w.Write([]byte(`{"status":"COMPLETED","id":"abc","ssn":"190001011234","name":"Test","givenName":"Test","surname":"User"}`))
	}))
	defer srv.Close()

	c := idkollen.NewClientBuilder("id", "secret").BaseURL(srv.URL).Build()
	status, err := c.BankIdSe().WaitForAuth(context.Background(), "abc", idkollen.PollOptions{})
	if err != nil {
		t.Fatal(err)
	}
	if _, ok := status.(*idkollen.BankIdSeCompleted); !ok {
		t.Fatalf("expected *BankIdSeCompleted, got %T", status)
	}
}
